platforms/allwinner-d1/core/src/lib.rs

```diff
@@ -260,14 +265,32 @@ impl D1 {
     }
 
     pub fn handle_panic(info: &PanicInfo) -> ! {
+        // Disable interrupts.
+        unsafe {
+            riscv::interrupt::disable();
+        }
+
+        // Avoid double panics.
+        static PANICKING: AtomicBool = AtomicBool::new(false);
+        if PANICKING.swap(true, Ordering::SeqCst) {
+            die();
+        }
+
+        // TODO(eliza): abort any in-flight DMAs.
+
         // Ugly but works
         let mut uart: Uart = unsafe { core::mem::transmute(()) };
 
         write!(&mut uart, "\r\n").ok();
```

Comment:
You might want a null BEFORE as well, to force the end of the previous frame.

Additionally, you also probably want to either busy wait until the DMAC transfer is complete, or hard-stop any DMAC transfers that are ongoing, before manually driving the panic, otherwise I dunno how pushing bytes into the fifo while the DMAC is also doing it would go.

Reply:
yeah, i left a TODO about killing DMAC transfers. i intend to do that in a follow-up but will have to read the data sheet first.